storage/s3.go

```diff
@@ -126,13 +126,72 @@ func (s *S3) Get(ctx context.Context, path string) (io.ReadCloser, error) {
 		Bucket: &s.cfg.Bucket,
 		Key:    &path,
 	}
+
 	out, err := s.s3client.GetObject(ctx, input)
 	if err != nil {
 		return nil, errors.Wrapf(err, "fail to get S3 object %v", path)
 	}
 	return out.Body, nil
 }
 
+// GetWithRetries function downloads the object from S3 and writes it to the
+// writer. It returns the number of bytes written to the writer.
+//
+// It implements a retry mechanism to handle if connection is closed before
+// the end of the download.
+func (s *S3) GetWithRetries(ctx context.Context, path string, writer io.Writer) (int64, error) {
+	path = fullPath(path)
+	log := logger.Get(ctx)
+	log.WithField("path", path).Info("Get object with retries")
+
+	size, err := s.Size(ctx, path)
+	if err != nil {
+		return -1, errors.Wrapf(err, "get size of S3 object %v", path)
+	}
+
+	var (
+		offset  int64
+		attempt = 1
+	)
+
+	for {
+		// Log if we are resuming a download
+		if offset != 0 {
+			log.Infof("Downloading from offset %d...", offset)
+		}
+
+		offsetSize := size - offset
+		rangeHeader := fmt.Sprintf("bytes=%d-%d", offset, size-1)
+		resp, err := s.s3client.GetObject(ctx, &s3.GetObjectInput{
+			Bucket: &s.cfg.Bucket,
+			Key:    &path,
+			Range:  aws.String(rangeHeader),
+		})
+		if err != nil {
+			if attempt >= s.retryPolicy.Attempts {
+				return -1, errors.Wrap(err, "get object after retries")
+			}
+			time.Sleep(s.retryPolicy.WaitDuration)
+			log.Infof("Get Object request error: %v, retrying", err)
```

Comment:
suggestion(non-blocking): errors should be logged as logger fields
```suggestion
			log.WithError(err).Info("Get Object request error, retrying")
```

Reply:
Ack